<!DOCTYPE html><html lang="en"><head><script async src="https://www.googletagmanager.com/gtag/js?id=UA-114897551-4"></script><script>window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'UA-114897551-4');
</script><script type="text/javascript" src="//platform-api.sharethis.com/js/sharethis.js#property=5ac2443d1fff98001395ab6c&amp;product=sticky-share-buttons" async="async"></script><title>CIS163-101: Try Catch Blocks | Coder Coacher - Coaching Coders</title><meta content="CIS163-101: Try Catch Blocks - All technical stuff in one place" name="description"><meta name="keywords" content="education, coding, programming, technology, nodejs, mongodb, software, computer science, engineering, teaching, coaching, coder, learning, java, kotlin, machine learning, AI, ML, tech talks, angular, javascript, js, typescript"><meta name="viewport" content="width=device-width, initial-scale=1.0"><link rel="stylesheet" href="/css/font.css"><link rel="stylesheet" href="/css/bootstrap.css"><link rel="stylesheet" href="/css/style.css"><link rel="stylesheet" href="/css/coder-coacher.css"></head><body><div class="container-fluid"><h1 class="site-title"><a href="/">Coder Coacher</a></h1><hr><h4 class="site-subtitle text-right">Coaching Coders</h4></div><div id="amzn-assoc-ad-99d6751e-2392-4004-ad16-73aa8385d9d0"></div><script async src="//z-na.amazon-adsystem.com/widgets/onejs?MarketPlace=US&amp;adInstanceId=99d6751e-2392-4004-ad16-73aa8385d9d0"></script><div class="post__breadcrumb"><div class="container"><ol class="breadcrumb"><li><a href="/">Coder Coacher</a></li><li><a href="/IAmTimCorey/">IAmTimCorey</a></li><li class="active">â¤µ</li></ol></div></div><h2 class="post__title"><b>CIS163-101: Try Catch Blocks</b></h2><h5 class="post__date">2013-10-08</h5><div class="container"><div class="video-responsive"><iframe width="560" height="315" src="https://www.youtube.com/embed/Nhq5m5UmJiA" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe></div><div class="post__text">in this video we are going to discuss
error handling more specifically we're
going to discuss try-catch okay here I
have a simple console application I have
just my static void main with just a few
lines of code what I'm doing here is
creating couple variables a string
variable called age text and integer
variable called age I then put out to
the console what is your age and that
same line I read in whatever the user
types I put this into the text variable
since read line returns a string I then
use the int dot parse to convert that
string into a number and I then report
that back to the console doing a
console.writeline so it should say your
age is and then whatever you typed in
here and finally I do my console dot
read to stop the execution of the
application okay let's see how this runs
I press Start what is your age 35 your
age is 35 pretty simple okay
let's see what happens though this time
if I type out 40 hit enter it pauses for
a minute thinks about it tries it and
that blows up and we get this nasty
exception handled box down here in this
case it's a format exception and if we
read down here it says additional
information input string was not in a
correct format what that's saying is I
expected a number and you gave me text
so there are ways around this the way we
went over in class was the tryparse
which tries to do the conversion and
fails gracefully in this video we're
going to discuss how to use try-catch
instead so there's a
waiting to do this first of all we could
do a control X here and cut this line
out and then type in try and use a
try-catch snippet so he's a tab tab and
then inside of our tribe a paste back
our code that's one way but there is
another way if we come back here and put
our code back in if we already have the
code we want inside try-catch in our
code simply highlight it right click and
say surround width and then type in TR Y
for the try snippet hit tab and it puts
the age equals in parse age text line
inside of the try okay makes a little
simpler if you already have code there
if you don't just hit try tab to Abin it
pops it up just fine
okay so what happens is when the code
starts reading it rate gets line 17 it
reads that in gets the line 19 and says
okay this is a try-catch so I'm going to
do everything inside of the curly braces
so try so it's going to in this case
just run line 21 if there's a problem in
line 21 where it throws an exception we
hit one of the catch exception catch
blocks okay and then it will run
everything inside the curly brace is
there if there are no exceptions in line
21 it skips down below the catch block
and continues processing so in this case
if we type in 35 into our age text it'll
parse just fine and keep going
however if we type in 40 that will throw
an exception this catch block has a type
exception which is the base exception
meaning it catches all types of
exceptions so anything in here
we'll be executed now in this case the
snippet adds in this throw what this
throw does is it rethreads the exception
the point of this is that if you are
deep into a program and you catch an
exception you may want to pass that
exception back up the chain but maybe
you want to do something with it first
may you want to reset a variable or
maybe you want to add information to the
exception stack that's where you do that
before you throw it ok but for our case
we don't want to three throw the
exception we want to stop the exception
from happening we wanted to stop we want
to not stop our application so let's go
ahead and do a free generic
console.writeline that says unexpected
exception has occurred and let's go
ahead and close that out all right so
let's run this again see what happens
your age is 35 and pops right back up
just like normal let's try it again and
this time we'll type in 40 an unexpected
exception has occurred or occurred your
age is zero well it's zero because we
never actually changed that age variable
after we initialized it because it
failed in the Tri okay so that's that's
pretty good but you know what maybe we
want to say we want to catch that
specific exception that was thrown well
we can add another catch block and this
time type out form an exception and our
credit brace is in and move that back in
the line now what's going to happen here
is because we've already seen that the
format exception is what gets thrown if
you type in 40 so when this throws that
exception it will jump down into this
catch block because it's the first one
that matches if it's a different type of
exception then it will jump down into
this catch block but it only runs one of
the cache blocks so let's write out
something more specific
you had not enter a valid number okay so
let's go ahead and run this again we
already know 35 works so let's type in
40 you did not enter a valid number
you're a 2-0 okay so what we see is that
we run the try and we throw that
exception it runs just this catch block
and it ignores this catch block okay you
can have as many catch blocks as you
want after your try the other thing you
can do is you can give this exception a
name of an instance name typically you
see a X being used okay and actually you
can give that same name to both
exceptions because they're inside the
scope of these curly braces okay that
e^x is not used outside of this scope
the benefit here is that you can do fun
stuff like this where you say
console.writeline e^x dot inner
exception or dot data or in this case
let's just do message what this will do
is we'll put the message on the console
as well so if we run this if I type in
40 you did not enter valid number but
then input string was not in the correct
format okay so that might be useful in
some circumstances where you might want
to put out a cleaner error message to
the console but you don't want to type
out every different type of message
notice the green squiggly is here
because I am not actually using that
variable so the variable does not have
to be include in fact it's recommend you
do not include it unless you're actually
going to use it somewhere inside your
catch block we haven't seen a need for
this in class yet but there is a case
where you could use a finally block
a finally block will execute after the
try and after the catch so for example
if you type in 35 into the age text the
try will execute and then the finally it
will execute if you type in 40 in the
age text the try will execute this catch
blocks will will execute and then this
finally will execute the purpose of the
finally block is to handle things like
closing a database connection or other
operations that need to happen no matter
what whether there's an exception or not
and that's what the finally is for in
most cases however try-catch is all you
need so that's a basic overview of
try-catch note that in a lot of
Microsoft's libraries you see this list
of exceptions you see that in this case
the parse method has three different
types of exceptions it could throw
argument null exception format exception
an overflow exception okay so these you
could create catch blocks for each of
those circumstances but at least you
need to know these are the possible
exceptions that will be thrown okay so
try everything inside runs and there's
an error the first catch block that
matches the exception gets executed and
then if you have an optional finally
that gets executed no matter what for a
try-catch block you can only have one
try and you must have one or more catch
blocks and that's all there is to know
about try catch</div></div><div class="container-fluid bottom-ad"><div id="amzn-assoc-ad-6a809dda-347a-4187-8a86-91faf94575da"></div><script async src="//z-na.amazon-adsystem.com/widgets/onejs?MarketPlace=US&amp;adInstanceId=6a809dda-347a-4187-8a86-91faf94575da"></script></div><div class="text-center">We are a participant in the Amazon Services LLC Associates Program, an affiliate advertising program designed to provide a means for us to earn fees by linking to Amazon.com and affiliated sites.</div><script>(function(w, d){
    var b = d.getElementsByTagName('body')[0];
    var s = d.createElement("script"); s.async = true;
    var v = !("IntersectionObserver" in w) ? "8.6.0" : "10.4.2";
    s.src = "https://cdnjs.cloudflare.com/ajax/libs/vanilla-lazyload/" + v + "/lazyload.min.js";
    w.lazyLoadOptions = {};
    b.appendChild(s);
}(window, document));</script></body></html>